### ISSUE TITLE
Events: spell out the difference between Shoes and Display events or remove them

### ISSUE THREAD

**Issue by noahgibbs** [scarpe-team/scarpe#128]
Right now there are Shoes events and Display events as separate namespaces. I intended them to be roughly "shoes events are model-side events like a button was pressed, while display events are about the display implementation." But of course, that's a really messy distinction. It's not even as simple as "shoes widgets send shoes events, display widgets send display events" currently, though I suppose it *could* be. But would that be useful?

Either we should figure out a useful distinction between them or the distinction should go away and there should just be one event type.

Reasons we might care:

* we have a remoted display service with a socket between model-side and display-side -- do all events need to cross that socket? Do we want to somehow distinguish between events that do and don't get sent across? Right now we send everything, which feels very inefficient. And a manual subscription service (tell the far side of the socket what to forward) would add complexity and timing bugs.
* would be nice not to just have "event spaghetti" everywhere. It's really nice to have obvious convenience functions like "send this event between a Shoes button and a webview button" in both directions, and really clear why certain paths would be really common, e.g. "JS event turns into display-to-shoes controller-type event." But right now it's just an unrestricted mess that's allowed to connect anywhere it wants. That will get bad, fast, if left to grow unchecked. Categories of "event paths" like "shoes sends property updates to display lib" and "display lib indicates user actions to shoes" help to prevent spaghetti. This structure also helps to answer questions like "do these events need to cross the socket, or can they just propagate locally?"

**Comment by noahgibbs** [comment]
Okay, more messing with tests is convincing me how this should work.

Specifically: it makes a lot of sense to have an "event box" - a specific place to subscribe to events which is *not* durable/portable across multiple Scarpe apps. Just having "subscribe to an event name on the singleton DisplayService" is too much event spaghetti.

Instead, it makes sense to have a place to subscribe to Shoes events based on a single Shoes/Scarpe app. So things like "app starts running, so send a run event" should happen there. And things like JS events should happen on an analogous WebviewDisplayService-only "event box", rather than being something that can propagate to Scarpe widgets. It's reasonable to have (e.g.) a WebviewEditBox cross that gap, though - it can subscribe to the display-side-only JS-related events, and then turn them (where appropriate) into Shoes events.

But *that* is probably the way to distinguish between Shoes and Display events, which means some events are currently miscategorised. For instance, the Scarpe::App should never send *any* Display events, and the "init" and "run" events it sends should be Shoes events, not Display events. A remoted WebviewRelayDisplayService should send only Shoes events across that gap, never display events, which exist only on the child display-service side.

This will require some sorting out, though, because again, right now some events are tagged incorrectly.

As a lovely side effect, testing gets more reasonable. I added a "full_reset!" method to reset all the event subscriptions on the display service, and I consider that to be a code smell. Instead, if you subscribe to events for the specific Scarpe app, it goes away when the app does. If you have seven Scarpe apps at the same time (usually a bad idea, but totally possible), they'd have seven separate "worlds" of Shoes events. If there's no display service, there would be no display events.

Basically, right now, having a singleton object and subbing everybody to it is a mess. Once this is sorted, the messes will be restricted to *inside* each Scarpe App and display service object, instead of a single giant mess across all of them. That's much more manageable. And then if we want to manage further, inside each one, we can do that.

**Comment by noahgibbs** [comment]
When I fix this, I need to look into heartbeat events, which suffer from the current somewhat-muddy setup. Specifically, we send a "heartbeat" display event from lib/shoes/app.rb's run method, but only if the display library does *not* take over run with its own event loop. And the local-WV display library does *not* send the same heartbeat event (I think?), instead using the JS heartbeat and control-interface heartbeat.

This is too many different heartbeats, not always available. And doing it differently per-display-service sucks.

(Yes, I wrote this code. It's still wrong and needs replacing.)

**Comment by noahgibbs** [comment]
In addition to heartbeats, things like "destroy" are funky right now. There's not a clear way to send a destroy signal from the display side, and there really should be. Existing display libraries that solve this are doing it by exiting the process, which isn't great and causing trouble for webview-relay, which *can't* easily exit the parent process.

**Comment by noahgibbs** [comment]
Starting to work on this, some things become clear.

"Display" events are probably an internal detail of the display service: the portable-to-different-display-servers parts of Scarpe shouldn't see them. It would be fine to use Scarpe event code for that, but not clear there's much benefit to doing so.

We don't need to have a modular object that can accept a bunch of subscriptions, I don't think. The big reason for "running a bunch of tests means it would be good to have subscriptions on an object so we can clean them up" is the idea of having more than one Scarpe app run in the same process, consecutively. But if we give up on multiple display services, or no display services, for a *single* app then that becomes a lot less useful. It becomes just part of mocked tests, and those shouldn't be expecting a "real" app object in any case. Any non-mocked test is going to have exactly one display service object, ever, which will be destroyed at the same time the app finishes.

So instead, I think it's okay that the DisplayService singleton object accepts all the subscriptions. It's going to run one app, ever. Needing/wanting the reset! method indicates that I'm misusing it, either by using it for mocking or trying to run more than one display service (consecutively) in the same process. Some display services could (practically, in concept) do so, but we shouldn't say that's possible in the API, because display services like Webview (and many others) don't permit it. So we should just say "don't do that" rather than trying to partially allow something we know is a terrible idea.

**Comment by noahgibbs** [comment]
I've gotten rid of display events - all the current events are shoes events, and should be.

**Comment by noahgibbs** [comment]
Also in #228 I think I sorted out the destroy event, including allowing it to be sent from the display service side. But I need to test that, and also sort out heartbeat.

**Comment by noahgibbs** [comment]
Enough of this is fixed that I'll close it and open a smaller issue with just the destroy/heartbeat remaining issues.
